File: UI/app/src/main/java/com/rikkei/training/ui/MainActivity.java
package com.rikkei.training.ui;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;

import android.os.Bundle;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {
    private RecyclerView recyclerLinear;
    private RecyclerView recyclerGrid;
    private RecyclerView recyclerStaggered;
    private Adapter adapter;
    private Adapter adapterGrid;
    private List<Image> data;
    private List<Image> dataGrid;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        initView();
        initData();
    }

    private void initData() {
        data = new ArrayList<>();
        data.add(new Image(R.drawable.ic_launcher_background));
        data.add(new Image(R.drawable.ic_launcher_background));
        data.add(new Image(R.drawable.ic_launcher_background));
        data.add(new Image(R.drawable.ic_launcher_background));
        data.add(new Image(R.drawable.ic_launcher_background));
        data.add(new Image(R.drawable.ic_launcher_background));
        data.add(new Image(R.drawable.ic_launcher_background));

        dataGrid = new ArrayList<>();
        dataGrid.add(new Image(R.drawable.image));
        dataGrid.add(new Image(R.drawable.image));
        dataGrid.add(new Image(R.drawable.image));
        dataGrid.add(new Image(R.drawable.image));
        dataGrid.add(new Image(R.drawable.image));
        dataGrid.add(new Image(R.drawable.image));
        dataGrid.add(new Image(R.drawable.image));
        dataGrid.add(new Image(R.drawable.image));
        dataGrid.add(new Image(R.drawable.image));
        dataGrid.add(new Image(R.drawable.image));

        adapter = new Adapter(this);
        adapter.setData(data);
        adapterGrid = new Adapter(this);
        adapterGrid.setData(dataGrid);
        recyclerLinear.setAdapter(adapter);
        recyclerGrid.setAdapter(adapterGrid);
        recyclerStaggered.setAdapter(adapter);
    }

    private void initView() {
        recyclerLinear = findViewById(R.id.recycler_linear);
        recyclerGrid = findViewById(R.id.recycler_grid);
        recyclerStaggered = findViewById(R.id.recycler_staggered);
    }
}

File: LiveDataKotlin/app/src/main/java/com/rikkei/training/livedatakotlin/ui/MainActivity.kt
package com.rikkei.training.livedatakotlin.ui

import android.app.Activity
import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Toast
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.RecyclerView
import com.rikkei.training.livedatakotlin.*
import com.rikkei.training.livedatakotlin.database.WordListAdapter
import com.rikkei.training.livedatakotlin.model.Word
import com.rikkei.training.livedatakotlin.viewmodel.WordViewModel
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {
    private lateinit var wordViewModel: WordViewModel

    companion object {
        const val newWordActivityRequestCode = 1
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        init()
    }

    private fun init() {
        fab.setOnClickListener {
            val intent = Intent(this, NewWordActivity::class.java)
            startActivityForResult(intent,
                newWordActivityRequestCode
            )
        }

        val recyclerView = findViewById<RecyclerView>(R.id.recyclerview)
        val adapter = WordListAdapter(this)
        recyclerView.adapter = adapter

        wordViewModel = ViewModelProviders.of(this).get(WordViewModel::class.java)
        wordViewModel.allWords.observe(this, Observer { words ->
            words?.let { adapter.setWords(it) }
        })
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == newWordActivityRequestCode && resultCode == Activity.RESULT_OK){
            data?.let {
                val word =
                    Word(it.getStringExtra(NewWordActivity.EXTRA_REPLY))
                wordViewModel.insert(word)
            }
        }else{
            Toast.makeText(applicationContext,
                R.string.empty_not_saved,Toast.LENGTH_SHORT).show()
        }
    }
}
